### PR TITLE
Feature: New SQL syntax WITH for Query and QueryBuilder

### PR DESCRIPTION
**Pull request by sartor**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Tests pass?   | ✔️

New MySQL version and PgSQL supports for WITH syntax as a part from standard.
https://dev.mysql.com/doc/refman/8.0/en/with.html
https://www.postgresql.org/docs/current/queries-with.html

Support for WITH prefix for Query and QueryBuilder with optional RECURSIVE modificator

### CODE REVIEW

File: framework/db/QueryBuilder.php [hunk 1]
```diff
@@ -260,6 +260,11 @@ public function build($query, $params = [])
             $sql = "($sql){$this->separator}$union";
         }
 
+        $with = $this->buildWithQueries($query->withQueries, $params);
+        if ($with !== '') {
+            $sql = "$with{$this->separator}$sql";
```
Comment (xepozz) [inline framework/db/QueryBuilder.php]:
```suggestion
            $sql = $with . $this->separator . $sql;
```

File: framework/db/sqlite/QueryBuilder.php [hunk 2]
```diff
@@ -520,6 +520,11 @@ public function build($query, $params = [])
             $sql = "$sql{$this->separator}$union";
         }
 
+        $with = $this->buildWithQueries($query->withQueries, $params);
+        if ($with !== '') {
+            $sql = "$with{$this->separator}$sql";
```
Comment (xepozz) [inline framework/db/sqlite/QueryBuilder.php]:
```suggestion
            $sql = $with . $this->separator . $sql;
```

File: framework/db/Query.php [hunk 3]
```diff
@@ -451,6 +461,7 @@ protected function queryScalar($selectExpression, $db)
             && empty($this->groupBy)
             && empty($this->having)
             && empty($this->union)
+            && empty($this->with)
```
Comment (xepozz) [inline framework/db/Query.php]: ???

File: framework/db/Query.php [hunk 4]
```diff
@@ -110,6 +110,16 @@ class Query extends Component implements QueryInterface, ExpressionInterface
...
+
+    /**
+     * @var array this is used to construct the WITH section in a SQL query.
+     * Each array element is an array of the following structure:
+     *
+     * - `query`: either a string or a [[Query]] object representing a query
+     * - `alias`: string, alias of query for further usage
+     * - `recursive`: boolean, whether it should be `WITH RECURSIVE` or `WITH`
+     */
+    public $withQueries;
```
Comment (xepozz) [inline framework/db/Query.php]: Why not just `$with`?

File: tests/framework/db/sqlite/QueryBuilderTest.php [hunk 5]
```diff
@@ -143,6 +143,37 @@ public function testBuildUnion()
         $this->assertEquals([], $queryParams);
     }
 
+    public function testBuildWith()
```
Comment (xepozz) [inline tests/framework/db/sqlite/QueryBuilderTest.php]: Please rename test methods to name of original methods with prefix `test`.
If testable method named as `buildWithQueries` then name test method `testBuildWithQueries`.

**Review by xepozz — CHANGES_REQUESTED** [review]

File: framework/db/Query.php (hunk 4 above)
Comment (sartor) [reply framework/db/Query.php]: because of same name property in ActiveQueryBuilder class

File: framework/db/QueryBuilder.php (hunk 1 above)
Comment (sartor) [reply framework/db/QueryBuilder.php]: I wrote it in a same way as union concatenation above

File: tests/framework/db/sqlite/QueryBuilderTest.php (hunk 5 above)
Comment (sartor) [reply tests/framework/db/sqlite/QueryBuilderTest.php]: Ok. Thank you for review

File: framework/db/QueryBuilder.php (hunk 1 above)
Comment (sartor) [reply framework/db/QueryBuilder.php]: I've benchmark both constructions: "$with{$this->separator}$sql" - is near 15% faster for same memory consumption

Comment (xepozz) [reply framework/db/QueryBuilder.php]: Oh my :smile: 
Ok, leave your version

File: framework/db/Query.php (hunk 3 above)
Comment (xepozz) [reply framework/db/Query.php]: What about this? property `$with` not exists in Query

Comment (sartor) [reply framework/db/Query.php]: I removed this condition

**Comment by sartor** [comment]
Is there something else I need to fix?

**Comment by samdark** [comment]
Does that work for SQLite, MSSQL, Oracle?

**Comment by sartor** [comment]
> Does that work for SQLite, MSSQL, Oracle?

It 100% work with PostgreSQL, MySQL and SQLite. Not sure about MS and Oracle, but from documentation they have exactly same syntax:
https://docs.microsoft.com/en-us/sql/t-sql/queries/with-common-table-expression-transact-sql?view=sql-server-ver15
https://www.geeksforgeeks.org/sql-with-clause/

**Comment by samdark** [comment]
@yiisoft/reviewers need opinions.

**Comment by terabytesoftw** [comment]
It should be merged too.

**Comment by xepozz** [comment]
I think it shouldn't be released because yii2 has feature freeze status

**Comment by sartor** [comment]
> 
> 
> I think it shouldn't be released because yii2 has feature freeze status

I will make same pull request to yii3 db package if this one is approved

**Comment by xepozz** [comment]
> > 
> > 
> > I think it shouldn't be released because yii2 has feature freeze status
> 
> I will make same pull request to yii3 db package if this one is approved

Sure, you can do it.

**Comment by terabytesoftw** [comment]
@sartor hi, already in yii3 this merge the commit, but you are welcome to collaborate, regarding this commit it should merge, because everything progresses, and it can be useful, it is a new function, it does not break bc, and it gives more compatibility with Mysql 8.0.

**Comment by samdark** [comment]
Thought more about it. The quality of the PR and the fact that it won't break anything allows us to add it to next release.

**Review by samdark — COMMENTED** [review]
Need the following:

- Entry for CHANGELOG
- Make sure that all relevant docs are updated

**Comment by sartor** [comment]
I'll add docs and comments today

File: framework/db/QueryBuilder.php [hunk 6]
```diff
@@ -1492,6 +1497,31 @@ public function buildUnion($unions, &$params)
         return trim($result);
     }
 
+    public function buildWithQueries($withs, &$params)
```
[the remark this answers is not in the document]
Comment (sartor) [reply framework/db/QueryBuilder.php:1507]: +

File: framework/db/Query.php [hunk 7]
```diff
@@ -1250,6 +1260,19 @@ public function union($sql, $all = false)
         return $this;
     }
 
+    /**
+     * Prepends a SQL statement using WITH syntax.
+     * @param string|Query $query the SQL statement to be appended using UNION
+     * @param string $alias query alias in WITH construction
+     * @param bool $recursive TRUE if using WITH RECURSIVE and FALSE if using WITH
+     * @return $this the query object itself
```
[the remark this answers is not in the document]
Comment (sartor) [reply framework/db/Query.php:1270]: done

File: framework/db/Query.php [hunk 8]
```diff
@@ -110,6 +110,16 @@ class Query extends Component implements QueryInterface, ExpressionInterface
...
      */
     public $union;
+
+    /**
+     * @var array this is used to construct the WITH section in a SQL query.
+     * Each array element is an array of the following structure:
+     *
+     * - `query`: either a string or a [[Query]] object representing a query
+     * - `alias`: string, alias of query for further usage
+     * - `recursive`: boolean, whether it should be `WITH RECURSIVE` or `WITH`
```
[the remark this answers is not in the document]
Comment (sartor) [reply framework/db/Query.php:120]: done

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Thank you!

**Comment by grachov** [comment]
Should this work with ActiveQuery also? As I see `withQueries` value is lost after calling `ActiveQuery::prepare`.

**Comment by sartor** [comment]
I didn't check it.

**Comment by grachov** [comment]
This does not sound serious :) I'll create a PR for this.

**Comment by sartor** [comment]
In my projects I never use ORM and Active queries. That's why I didn't cover it. Thank you for contributing!

**Pull request merged**